### PR TITLE
feat: Make it possible to be abstract over bounding volumes

### PR DESCRIPTION
**Pull request by Rhuagh**
This change is [Reviewable](https://reviewable.io/reviews/kvark/collision-rs/59)

### CODE REVIEW

File: src/algorithm/broad_phase/sweep_prune.rs [hunk 1]
```diff
@@ -83,11 +82,11 @@ where
         }
 
         shapes.sort_by(|a, b| {
-            let cmp_min = a.get_bound().min()[self.sweep_axis]
-                .partial_cmp(&b.get_bound().min()[self.sweep_axis]);
+            let cmp_min = a.get_bound().min_extent()[self.sweep_axis]
```
Comment (kvark) [inline src/algorithm/broad_phase/sweep_prune.rs]: let's rename `get_bound()` -> `bound()`, which highlights the fact it's just an accessor. If it's not, we should only call it once per iteration here.

File: src/algorithm/broad_phase/sweep_prune.rs [hunk 2]
```diff
@@ -137,14 +136,16 @@ where
...
+
+    use BoundingVolume;
+    use cgmath::{BaseFloat, Point2, Point3, Vector2, Vector3};
     use cgmath::prelude::*;
 
     /// Trait for variance calculation in sweep and prune algorithm
     pub trait Variance {
         /// Point type
-        type Bound: Aabb;
+        type Bound: BoundingVolume;
```
Comment (kvark) [inline src/algorithm/broad_phase/sweep_prune.rs]: any reason the trait is not simply called `Bound`?

File: src/dbvt/wrapped.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    B: BoundingVolume,
+    <B::Point as EuclideanSpace>::Diff: Debug,
+{
+    /// The value
+    pub value: V,
+
+    /// The bounding volume
+    pub bound: B,
+
+    fat_factor: <B::Point as EuclideanSpace>::Diff,
```
Comment (kvark) [inline src/dbvt/wrapped.rs]: fat factor?

File: src/dbvt/wrapped.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+impl<'a, V, H> From<(V, &'a H)> for TreeValueWrapped<V, H::Bound>
+where
+    H: HasBound,
+    H::Bound: Clone,
+    <<H::Bound as BoundingVolume>::Point as EuclideanSpace>::Diff: Debug + Zero,
+{
+    fn from((value, f): (V, &H)) -> Self {
+        Self::new(value, f.get_bound(), <<H::Bound as BoundingVolume>::Point as EuclideanSpace>::Diff::zero())
+    }
+}
```
Comment (kvark) [inline src/dbvt/wrapped.rs]: nit: newline

File: src/primitive/capsule.rs [hunk 5]
```diff
@@ -45,17 +59,46 @@ where
     }
 }
 
+impl<'a, S> From<&'a Capsule<S>> for Aabb3<S>
```
Comment (kvark) [inline src/primitive/capsule.rs]: I don't think converting from reference is idiomatic. It's supposed to have `move` semantics, hence the auto-derive of `Into` trait.

File: src/primitive/capsule.rs [hunk 6]
```diff
@@ -45,17 +59,46 @@ where
...
+where
+    S: BaseFloat,
+{
+    fn from(capsule: &Capsule<S>) -> Aabb3<S> {
```
Comment (kvark) [inline src/primitive/capsule.rs]: nit: use `Self`

**Comment by Rhuagh** [comment]
Review status: 0 of 19 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/algorithm/broad_phase/sweep_prune.rs, line 148 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuPP6cXEpH6L48W-r1-148:-L1lkMt7l7KcLQMGp2pf:b-t0ft6o) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/algorithm/broad_phase/sweep_prune.rs#L148)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

any reason the trait is not simply called `Bound`?
</blockquote></details>

Because of https://github.com/kvark/collision-rs/blob/master/src/bound.rs

---

*[src/dbvt/wrapped.rs, line 21 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuQp4anoqpleFGK-r1-21:-L1lkQnlokS-ZJ0UGFqb:b8g5rp9) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/dbvt/wrapped.rs#L21)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

fat factor?
</blockquote></details>

Used to create a bigger bounding volume for use inside the DBVT. It's basically a way of allowing a small amount of movement of volumes without rebalancing the tree.

---

*[src/primitive/capsule.rs, line 62 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuPP6cXEpH6L48Z-r1-62:-L1lkt2JKL2bAnd-RgUL:b-he84cq) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/primitive/capsule.rs#L62)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

I don't think converting from reference is idiomatic. It's supposed to have `move` semantics, hence the auto-derive of `Into` trait.
</blockquote></details>

I know this, I just felt this was better than using a self defined trait for computing the bounding volume of primitives.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by Rhuagh** [comment]
Review status: 0 of 19 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/algorithm/broad_phase/sweep_prune.rs, line 85 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuPP6cXEpH6L48W-r1-85:-L1llvHI7MmuiTJlGR5I:b-896fix) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/algorithm/broad_phase/sweep_prune.rs#L85)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

let's rename `get_bound()` -> `bound()`, which highlights the fact it's just an accessor. If it's not, we should only call it once per iteration here.
</blockquote></details>

Done.

---

*[src/dbvt/wrapped.rs, line 84 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuQp4anoqpleFGK-r1-84:-L1lltapH6nPasSw5wtf:b-896fix) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/dbvt/wrapped.rs#L84)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: newline
</blockquote></details>

Done.

---

*[src/primitive/capsule.rs, line 66 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuPP6cXEpH6L48Z-r1-66:-L1lluagPktBNTUK7FGO:b-896fix) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/primitive/capsule.rs#L66)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: use `Self`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by Rhuagh** [comment]
Review status: 0 of 19 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/primitive/capsule.rs, line 62 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuPP6cXEpH6L48Z-r1-62:-L1ls8r5eTwknMcMPPIk:b-w0xpip) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/primitive/capsule.rs#L62)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

I know this, I just felt this was better than using a self defined trait for computing the bounding volume of primitives.
</blockquote></details>

You know what, I change my mind. I'll make it a locally defined trait instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by Rhuagh** [comment]
Review status: 0 of 19 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/primitive/capsule.rs, line 62 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuPP6cXEpH6L48Z-r1-62:-L1lxZnGZwwzC-TSIDm8:b-896fix) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/primitive/capsule.rs#L62)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

You know what, I change my mind. I'll make it a locally defined trait instead.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by kvark** [comment]
Great stuff right there, and we are almost done with the review! Just a few more concerns to answer ;)

---

Reviewed 3 of 19 files at r1, 16 of 16 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[src/traits.rs, line 62 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1oancEVsanoYGLqhyu:-L1oancEVsanoYGLqhyv:b-30lcp4) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/traits.rs#L62)):*
> ```Rust
>     /// Create a new bounding volume extended by the given amount
>     fn with_margin(&self, add: <Self::Point as EuclideanSpace>::Diff) -> Self;
>     /// Apply an arbitrary transform to the bounding volume.
> ```

extreme nit: let's try to be consistent about the '.' at the end of doc comments

---

*[src/traits.rs, line 138 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1obOqeFQJOX5DRg5uX:-L1obOqeFQJOX5DRg5uY:b-u4iqgz) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/traits.rs#L138)):*
> ```Rust
> 
>     /// Intersection test for transformed self
>     fn intersection_transformed<T>(&self, _: &RHS, _: &T) -> Option<Self::Result>
> ```

why do we need this change?

---

*[src/traits.rs, line 144 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1ob_shIPMy-TprvROF:-L1ob_shIPMy-TprvROG:b-cf057d) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/traits.rs#L144)):*
> ```Rust
> 
> /// Marker trait for a collision primitive.
> pub trait Primitive: Clone + SupportFunction {}
> ```

Is support function useful outside of primitives? If not, can we just move all the contents into `Primitive` and end up with less traits?

---

*[src/algorithm/broad_phase/sweep_prune.rs, line 148 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuPP6cXEpH6L48W-r1-148:-L1oaMYNW9jNqqbYrnI3:b2nhghv) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/algorithm/broad_phase/sweep_prune.rs#L148)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Because of https://github.com/kvark/collision-rs/blob/master/src/bound.rs
</blockquote></details>

Hmm

---

*[src/dbvt/mod.rs, line 126 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1ocHhRKuwTQhwtf3Gr:-L1ocHhTy0NE4CEy9PY6:bwcp4vb) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/mod.rs#L126)):*
> ```Rust
> }
> 
> impl<T> HasBound for (usize, T)
> ```

this is a bit unclear. Can we have a newtype of something (even a typedef?) for this pair, explaining what it means?

---

*[src/dbvt/wrapped.rs, line 21 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuQp4anoqpleFGK-r1-21:-L1oaPzUGyoinrWbiYS7:beglzq6) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/dbvt/wrapped.rs#L21)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Used to create a bigger bounding volume for use inside the DBVT. It's basically a way of allowing a small amount of movement of volumes without rebalancing the tree.
</blockquote></details>

wonderful, is this documented in the code somewhere?

---

*[src/dbvt/wrapped.rs, line 32 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1ocihEtGgIy27GOePQ:-L1ocihFcr4at9D9qoA8:b-3eu64m) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/wrapped.rs#L32)):*
> ```Rust
>     pub fn new(value: V, bound: &B, fat_factor: <B::Point as EuclideanSpace>::Diff) -> Self
>     where
>         B: Clone,
> ```

we can move `Clone` bound up in the `impl` section. Not sure why we can't just accept `B` by move here, shoving the responsibility to clone it to the caller?

---

*[src/dbvt/wrapped.rs, line 54 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1od6axwWpUU_cV-5zZ:-L1od6ay6ior-86Yws4L:b-5o32jp) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/wrapped.rs#L54)):*
> ```Rust
>     }
> 
>     fn fat_bound(&self) -> Self::Bound {
> ```

and here we need to reflect somewhat in the name that it's not an accessor. Maybe `get_bound_with__margin`?

---

*[src/dbvt/wrapped.rs, line 71 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1odU-GdFWhoEhMIEi1:-L1odU-HP4k4cnSWGy-q:b-jzril1) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/wrapped.rs#L71)):*
> ```Rust
> }
> 
> impl<'a, V, H> From<(V, &'a H)> for TreeValueWrapped<V, H::Bound>
> ```

is it possible to just have this method as a constructor in the `impl TreeValueWrapped`? Similar to another case, converting from a reference is not idiomatic...

---

*[src/volume/sphere.rs, line 27 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1ofFjwQfllJJgccArv:-L1ofFjysO-6f3VLHPDY:bo0dbg7) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/volume/sphere.rs#L27)):*
> ```Rust
>     fn min_extent(&self) -> Point3<S> {
>         Point3::new(
>             self.center.x - self.radius,
> ```

nit: could be `self.center - Vector3::from_value(self.radius)`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59#-:-L1ofp9r5Y2OIi6I3hX1:b-7s0q4n)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[src/traits.rs, line 138 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1obOqeFQJOX5DRg5uX:-L1qnUuvv70PMhhN4s6u:b-7rwose) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/traits.rs#L138)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

why do we need this change?
</blockquote></details>

Compiler complains about it, saying unnamed parameters in traits will be deprecated soon

---

*[src/dbvt/wrapped.rs, line 21 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuQp4anoqpleFGK-r1-21:-L1qnyuocOzyD4qjV1tT:bcymyd0) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/dbvt/wrapped.rs#L21)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

wonderful, is this documented in the code somewhere?
</blockquote></details>

On TreeValue I  believe

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by kvark** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[src/traits.rs, line 138 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1obOqeFQJOX5DRg5uX:-L1r35R-plptwuKQA5cg:b6tsd6q) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/traits.rs#L138)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Compiler complains about it, saying unnamed parameters in traits will be deprecated soon
</blockquote></details>

That sucks :( I really like the old name-less syntax

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[src/traits.rs, line 144 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1ob_shIPMy-TprvROF:-L1ra84Ef5rxkED-CuJq:b-j9lde9) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/traits.rs#L144)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

Is support function useful outside of primitives? If not, can we just move all the contents into `Primitive` and end up with less traits?
</blockquote></details>

Remnants of the old setup which required aabb bound. Ill check if consolidation make sense

---

*[src/algorithm/broad_phase/sweep_prune.rs, line 148 at r1](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1dsRuPP6cXEpH6L48W-r1-148:-L1raQ0NupOflJOXEBoV:b-syigrr) ([raw file](https://github.com/kvark/collision-rs/blob/07c6d03df4e75102f43dc8c8326e33f621d16838/src/algorithm/broad_phase/sweep_prune.rs#L148)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

Hmm
</blockquote></details>

We can change yhe name, it would just be abreaking change

---

*[src/dbvt/mod.rs, line 126 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1ocHhRKuwTQhwtf3Gr:-L1rac6a3JQ6_LRPiNf5:bwdqa0z) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/mod.rs#L126)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

this is a bit unclear. Can we have a newtype of something (even a typedef?) for this pair, explaining what it means?
</blockquote></details>

That is the internal storage format in the tree. This impl is here to make it possible to run broad phases durectly on the tree data

---

*[src/dbvt/wrapped.rs, line 32 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1ocihEtGgIy27GOePQ:-L1raygrM1ms29m3uRjO:b2fba2e) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/wrapped.rs#L32)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

we can move `Clone` bound up in the `impl` section. Not sure why we can't just accept `B` by move here, shoving the responsibility to clone it to the caller?
</blockquote></details>

Remnant of when this used to take HasAabb

---

*[src/dbvt/wrapped.rs, line 54 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1od6axwWpUU_cV-5zZ:-L1rbAYzipym2eaffoMt:b-obo1c5) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/wrapped.rs#L54)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

and here we need to reflect somewhat in the name that it's not an accessor. Maybe `get_bound_with__margin`?
</blockquote></details>

Yeah, thats a good idea

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[src/dbvt/wrapped.rs, line 71 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1odU-GdFWhoEhMIEi1:-L1rbYdYb3rUzUxe_kqw:b-7hq8zp) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/wrapped.rs#L71)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

is it possible to just have this method as a constructor in the `impl TreeValueWrapped`? Similar to another case, converting from a reference is not idiomatic...
</blockquote></details>

Hmm, that would make it impossible to use in a generic scope, i.e. rhusics

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by kvark** [comment]
@Rhuagh anything you are planning to do here, or ready to ship?

**Comment by Rhuagh** [comment]
I plan to update this

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/traits.rs, line 62 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1oancEVsanoYGLqhyu:-L1vQ-Qu7iAVsfgCBYgh:b-896fix) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/traits.rs#L62)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

extreme nit: let's try to be consistent about the '.' at the end of doc comments
</blockquote></details>

Done.

---

*[src/volume/sphere.rs, line 27 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1ofFjwQfllJJgccArv:-L1vQ8AjVb6Mc_h46Ric:b-896fix) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/volume/sphere.rs#L27)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

nit: could be `self.center - Vector3::from_value(self.radius)`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/dbvt/wrapped.rs, line 71 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1odU-GdFWhoEhMIEi1:-L1vSfLkn3rwpNs2t3qb:bseemuh) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/wrapped.rs#L71)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Hmm, that would make it impossible to use in a generic scope, i.e. rhusics
</blockquote></details>

The problem is that to remove this From, I would have to either create another new trait just for this, or I would have to put this method on TreeValue, but that would make TreeValue have a constructor which would only be useful for TreeValueWrapped and no other implementors of TreeValue. I'd rather do this non-idiomatic From instead to not confuse people with weird APIs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by Rhuagh** [comment]
Addressed review comments and squashed. Ready now if there are no more issues. Created a few followup issues.

**Comment by Rhuagh** [comment]
Review status: 15 of 20 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/dbvt/wrapped.rs, line 71 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1odU-GdFWhoEhMIEi1:-L1vmERum01adfn1PkIm:bqe2d3e) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/wrapped.rs#L71)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

The problem is that to remove this From, I would have to either create another new trait just for this, or I would have to put this method on TreeValue, but that would make TreeValue have a constructor which would only be useful for TreeValueWrapped and no other implementors of TreeValue. I'd rather do this non-idiomatic From instead to not confuse people with weird APIs.
</blockquote></details>

Actually, I figured out a way to do it. I'll update soon

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by Rhuagh** [comment]
Review status: 15 of 20 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/dbvt/wrapped.rs, line 71 at r3](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1odU-GdFWhoEhMIEi1:-L1vnXUqwOPJNsPUjfr5:b-896fix) ([raw file](https://github.com/kvark/collision-rs/blob/21fd4dd820e052868a5b386f80bb9c7b21f68bdd/src/dbvt/wrapped.rs#L71)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Actually, I figured out a way to do it. I'll update soon
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by kvark** [comment]
Looks good! Last few notes (really, really!) :)

---

Reviewed 6 of 6 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/dbvt/wrapped.rs, line 68 at r4](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1wosHsHicYhmg-BALW:-L1wosHtWrSIvCmvdgNX:b-w1ayhe) ([raw file](https://github.com/kvark/collision-rs/blob/9c52b54d9dffa119dd9e2bbb65056a8ac00eb63b/src/dbvt/wrapped.rs#L68)):*
> ```Rust
> }
> 
> impl<V, B> From<(V, B, <<B as BoundingVolume>::Point as EuclideanSpace>::Diff)>
> ```

I wonder if we can have an extra `P: EuclideanSpace` generic parameter just to reduce the signatures here

---

*[src/volume/sphere.rs, line 26 at r4](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1wp06YKmh3S8Zm3ACS:-L1wp06YKmh3S8Zm3ACT:bbkf0po) ([raw file](https://github.com/kvark/collision-rs/blob/9c52b54d9dffa119dd9e2bbb65056a8ac00eb63b/src/volume/sphere.rs#L26)):*
> ```Rust
> 
>     fn min_extent(&self) -> Point3<S> {
>         self.center + -Vector3::from_value(self.radius)
> ```

why not `self.center + Vector3::from_value(-self.radius)`? may be faster

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59#-:-L1wpDNF_rPBAiGFEynp:bpawqxl)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/dbvt/wrapped.rs, line 68 at r4](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1wosHsHicYhmg-BALW:-L1wrZDVvjIULA80jF0Y:b-896fix) ([raw file](https://github.com/kvark/collision-rs/blob/9c52b54d9dffa119dd9e2bbb65056a8ac00eb63b/src/dbvt/wrapped.rs#L68)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

I wonder if we can have an extra `P: EuclideanSpace` generic parameter just to reduce the signatures here
</blockquote></details>

Done.

---

*[src/volume/sphere.rs, line 26 at r4](https://reviewable.io:443/reviews/kvark/collision-rs/59#-L1wp06YKmh3S8Zm3ACS:-L1wrZo65H-qvMTekXyM:b-896fix) ([raw file](https://github.com/kvark/collision-rs/blob/9c52b54d9dffa119dd9e2bbb65056a8ac00eb63b/src/volume/sphere.rs#L26)):*
<details><summary><i>Previously, kvark (Dzmitry Malyshau) wrote…</i></summary><blockquote>

why not `self.center + Vector3::from_value(-self.radius)`? may be faster
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kvark/collision-rs/59)*

**Comment by kvark** [comment]
This moment has come! Nice work 👍 
bors r+

**Pull request merged**